spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java

```diff
@@ -0,0 +1,275 @@
+
+    // create 2 files under non-partitioned table
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+
+    List<Object[]> df1 = sql("SELECT * FROM %s ", tableName);
+
+    // set the min-input-files = 2, instead of default 5 to allow compaction
+    List<Object[]> output = sql(
+            "CALL %s.system.rewrite_data_files(table => '%s', options => map('min-input-files','2'))",
+            catalogName, tableIdent);
+
+    assertEquals("Action should rewrite 2 data files and add 1 data files",
+            ImmutableList.of(row(2, 1)),
+            output);
+
+    List<Object[]> df2 = sql("SELECT * FROM %s ", tableName);
+
+    Assert.assertEquals("data count should match", df1.size(), df2.size());
+  }
+
+  @Test
+  public void testRewriteDataFilesWithSortStrategy() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg", tableName);
+
+    // create 2 files under non-partitioned table
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+
+    List<Object[]> df1 = sql("SELECT * FROM %s ", tableName);
+
+    // set the min-input-files = 2, instead of default 5 to allow compaction
+    // set sort_order = id DESC LAST
+    List<Object[]> output = sql(
+            "CALL %s.system.rewrite_data_files(table => '%s', options => map('min-input-files','2'), " +
+                    "strategy => 'sort', sort_order => 'id DESC NULLS_LAST')",
+            catalogName, tableIdent);
+
+    assertEquals("Action should rewrite 2 data files and add 1 data files",
+            ImmutableList.of(row(2, 1)),
+            output);
+
+    List<Object[]> df2 = sql("SELECT * FROM %s ", tableName);
+
+    Assert.assertEquals("data count should match", df1.size(), df2.size());
+  }
+
+  @Test
+  public void testRewriteDataFilesWithFilter() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg", tableName);
+
+    // create 10 files under non-partitioned table
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+
+    List<Object[]> df1 = sql("SELECT * FROM %s ", tableName);
+
+    // select only 5 files for compaction (files having id = 1)
+    List<Object[]> output = sql(
+            "CALL %s.system.rewrite_data_files(table => '%s'," +
+                    " where => 'id = 1 and data is not null')", catalogName, tableIdent);
```

Comment:
I think this will actually rewrite all data files which *may* contain 1. Since the planning portion is done just on the column metrics and it wouldn't know whether 1 actually is used unless it's a partition value

Reply:
Agree. May be better to restrict to only partition spec columns.

Reply:
https://github.com/apache/iceberg/pull/3375/files#r748676338